File: 79-word-search/79-word-search.cpp
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size();
        int n = board[0].size();
        for (int i = 0;i < m;i++) {
            for (int j = 0;j < n;j++) {
                if (board[i][j] == word[0]) {
                    vector<vector<bool>> seen(m, vector<bool>(n, false));
                    if (dfs(board, word,i,j,m,n, seen)) return true;
                }
            }
        }
        return false;
    }
private:
    bool dfs(vector<vector<char>> &board, string &word, int i, int j, int m, int n, vector<vector<bool>> &seen) {
        if (word.size() == 0) return true;
        if (i >= m || j >= n || i < 0 || j < 0 || board[i][j] != word[0] || seen[i][j]) return false;
        seen[i][j] = true;
        string rem = word.substr(1);
        bool left = dfs(board, rem, i-1, j, m,n, seen);
        bool right = dfs(board, rem, i+1, j, m,n, seen);
        bool up = dfs(board, rem, i, j-1, m,n, seen);
        bool down = dfs(board, rem, i, j+1, m,n, seen);
        seen[i][j] = false;
        return left || right || up || down;
        
    }
};

File: 79-word-search/79-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        ROW, COL = len(board), len(board[0])
        dx = [0,0,-1,1]
        dy = [-1,1,0,0]
        def helper(r, c, i):
            if i == len(word):
                return True
            if (r < 0 or
               c < 0 or
               r >= ROW or
               c >= COL or
               word[i] != board[r][c]):
                return False
            tmp = board[r][c]
            board[r][c] = '#'
            for x in range(4):
                if helper(r+dx[x], c+dy[x], i+1):
                    return True 
            board[r][c] = tmp
            return False
        
        for i in range(ROW):
            for j in range(COL):
                if board[i][j] == word[0]:
                    if helper(i,j,0):
                        return True
        return False